clippy_lints/src/methods/manual_pattern_char_comparison.rs

```diff
@@ -0,0 +1,148 @@
+use std::cmp::Ordering;
+use std::ops::ControlFlow;
+
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::visitors::{for_each_expr, Descend};
+use hir::{BinOpKind, ExprKind, PatKind, PrimTy, QPath, TyKind};
+use rustc_ast::LitKind;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::def::Res;
+use rustc_lint::LateContext;
+use rustc_span::symbol::Symbol;
+
+use super::MANUAL_PATTERN_CHAR_COMPARISON;
+
+const PATTERN_METHODS: [(&str, usize); 22] = [
+    ("contains", 0),
+    ("starts_with", 0),
+    ("ends_with", 0),
+    ("find", 0),
+    ("rfind", 0),
+    ("split", 0),
+    ("split_inclusive", 0),
+    ("rsplit", 0),
+    ("split_terminator", 0),
+    ("rsplit_terminator", 0),
+    ("splitn", 1),
+    ("rsplitn", 1),
+    ("split_once", 0),
+    ("rsplit_once", 0),
+    ("matches", 0),
+    ("rmatches", 0),
+    ("match_indices", 0),
+    ("rmatch_indices", 0),
+    ("trim_start_matches", 0),
+    ("trim_end_matches", 0),
+    ("replace", 0),
+    ("replacen", 0),
+];
+
+pub(super) fn check(
+    cx: &LateContext<'_>,
+    _expr: &hir::Expr<'_>,
+    method_name: Symbol,
+    _receiver: &hir::Expr<'_>,
+    args: &[hir::Expr<'_>],
+) {
+    for &(method, pos) in &PATTERN_METHODS {
+        if method_name.as_str() == method
+            && args.len() > pos
+            && let arg = args[pos]
+            && let ExprKind::Closure(closure) = arg.kind
+            && closure.fn_decl.inputs.len() == 1
+            && let TyKind::Path(QPath::Resolved(_, c_arg_type)) = closure.fn_decl.inputs[0].kind
+            && c_arg_type.res == Res::PrimTy(PrimTy::Char)
```

Comment:
This part prevents e.g. `.find(|c| ...)` without an explicit type `c: char` being linted, a direct way to check the type would be  `cx.typeck_results().pat_ty(body.params[0].pat)`

However since a closure `Pattern` always takes a `char` so we can just remove this check 

```suggestion
```

Reply:
The problem that I have is that my lint isn't restricted to methods on string and so if I remove these lines it trigger [here](https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/search_is_some_fixable_none.rs#L57) which is a closure with an `usize` (following what I understand from the code and the debug I made). Is there a way to make sure my lint is triggered on string/str in a proper way or is it better to use the `cx.typeck_results().pat_ty(body.params[0].pat)` you proposed in case there is closure with a character in others function ?

Reply:
`single_char_patterns` checks that the receiver is a string so you'd inherit that when merging them

https://github.com/rust-lang/rust-clippy/blob/5aae5f6ae6b4ba0e6ddadea43e6a2bcadcaa2328/clippy_lints/src/methods/single_char_pattern.rs#L45-L46

That's better than checking with `pat_ty` because we don't want to suggest to replace say `"abc".chars().find(|c| c == 'a')` with `"abc".chars.find('a')`. It would be good to add that as a test case

Reply:
`Iterator::find` gives you a reference to the item so that example doesn't work, but a test such as

```rust
struct NotStr;

impl NotStr {
    fn find(&self, _: impl FnMut(char) -> bool) {}
}
```

would still be good

Reply:
I have added the test !